### ISSUE TITLE
Still repeated runs gives "DEBUG: Checksum too old for [file] ...

### ISSUE THREAD

**Issue by xxargs** [sahib/rmlint#381]
I compare very large sets of file between snapshot and master (on btrfs) before delete snapshot as last check. Size is around 1-2  TB and million of files each time - some of them is very large (disk images) and time consuming to hashed every time, so use xattr for caching of hash seems very attractive to speed up things (include already done xattr hashes in master following in taken snapshot from master later)    

Ubuntu have to old version of rmlint in apt-get  package system and missing feature,  so i get from github, compile it and hit in case #380 almost immediate. 

- i found out itself something is fishy with  'rm_xattr_read_hash' and more specific "if(FLOAT_SIGN_DIFF(g_ascii_strtod(mtime_buf, NULL), file->mtime, MTIME_TOL) == 0)"

search on github and found case #380 and adjust above to 
"if(FLOAT_SIGN_DIFF(g_ascii_strtod(mtime_buf, NULL), file->mtime, MTIME_TOL) != 0)"

compile and try again - now seems mostly passed as expected, but still hit some cluster of file (often very large file...) not will compare as same  even if 'xattr - l' on both files give exact same hash and timestamp in different snapshot.

Problem seems related to if timestamp are pure even integer value or have non zero decimal part

ie. if user.rmlint.blake2b.mtime: 1537112261,503916979 (not working) or 

user.rmlint.blake2b.mtime: 1423255823,000000000 (working)

I try GDB in emacs but is hard to trace depend works spread out on different thread and every time clear out xattr-value  on file if time not are compare,  but i possible make  hit of trouble point down here: 

268	    if(FLOAT_SIGN_DIFF(g_ascii_strtod(mtime_buf, NULL), file->mtime, MTIME_TOL) != 0) {
(gdb) print file->mtime
$7 = 1536619830.5669794
(gdb) print mtime_buf
$8 = "1536619830,566979408", '\000' <repeats 43 times>

I not experienced programmer (almost newbee) nor  familar with 'FLOAT_SIGN_DIFF' inner working - but seems trouble with rounding behaviour and/or truncate of decimal part in case of not even integers value timestamp  feeded from different source and why is not compare

**Comment by xxargs** [comment]
this i google translate from my language and qualiy  not the best... 

---

Have looked in the source code and see two things that need to be looked through

1: time resolution on timestamp becomes extremely sensitive as soon as they have decimal values ​​in themselves, an error of 0.00000001 in value is enough to make the difference between reusing an existing hash and having to re-hash the entire file again

As long as the timestamp is on a second basis, without decimal part, problem is not noticed, but today the time resolution is down to nanoseconds in several file systems.

One of the problems is that the 'mtime' have resolution is 7 decimal places but after conversion to ASCII is written with 9 decimal places in  timestamp, which is stored as xattr for the file.

When the value read at the next rmlint run, the ASCII timestamp is converted back to a double again with a lot of conversion artifacts from type 6-7 decimal and in later steps is compared with the file's mtime yet again with 7 decimal places resolution in a comparison requires exact numerical consistency down to the 8th decimal (defult on "MTIME_TOL"), to reuse the stored hash value instead of being forced to create a new hash.

The second thing seems to be related to decimal point and decimal comma respectively

I live in a country where use decimal comma , which causes a lot of mess
and in this case the function "g_ascii_strtod (mtime_buf, NULL)" in xattr.c does not seem to load the ASCII values ​​according to the 'locale' rules which they should do and as a result the timestamp ASCII values ​​retrieved from the file do not is loaded correctly when the string contains ","

i.e. ASCII timestamp like 1526146728,726771593 from xattr timestamp is loaded as 1526146728.00000000 .... without giving any errors and when compared to the file's 'mtime' 1526146728,7267715 for validity it will never match and the saved hash value will never be used for file comparisons.  in order for a match to match, it has to match up to the eighth decimal ...

'locale' seems to work when the program prints the ASCII values for 'mtime'  ​​and then prints with decimal comma in my case, but the reading seems to have problems and the question is whether locale 'reaches all parts of the programs when it runs such as. threaded processes

**Comment by sahib** [comment]
Hello @xxargs,

you're right in both cases. The issue was due to the limited precision and often cut off due to local issues. I pushed a fix in bbc896f0dddb899d3321a4d5a0e2cbb96b18d043 - please give it a try.

**Comment by xxargs** [comment]
I don't test your code yet,  

but I find more ting give impact for country with decimal comma.

for writing out timestamp to string  uses 'snprintf' to make strings from double and this honor 'locale'

but reading strings later uses 'g_ascii_strtod' and is _not_ honor 'locale' and wanted decimal point in string and readings erroneously (but not give an error)  in value if reading ',' in string  (all decimal setting to zero in most cases)  

But using 'g_strod' instead for 'g_ascii_strtod' seems honor 'locale' again  and routine start at working as expected.

**Comment by sahib** [comment]
That's why ``g_ascii_dtostr`` was used instead of ``snprintf`` now. It always uses a dot as decimal separator. ``g_ascii_strtod`` is actually locale-insensitive, which is good since the xattr values are not targeted for human consumption.

**Comment by xxargs** [comment]
I see, seems missing actual function was changed i reading diff-files.

But unfortunately,  I found more strange thing, is not tracked down yet and not always happen -  but for large files, say > 30 GByte  each seems checksum written to xattribute for files in directory long before is possible can be calculated (based of time needs for hashing for each file ), and also different value for exact same copy of files on different places  - and worse, this cannot fix it with re-run again with -U, -C and -xattr* argument depend  rmlint find files differ from reading of xattr in early stage (and seems have highest priority) and judge them as different and put out them from hashing later. 

This happen only if try using with one or more of  -U, -C, xattr* functions active, without them rmlint running as expected 

Is also not working to use --xattr=write to force rewritten hashes or using --xattr=clear to clear out xattr post ie.  already out of scope in early stage sorting and nothing happen...  

if using: 
"find  -type f -print0 | xargs -0 -r -I {}  xattr -d user.rmlint.blake2b.mtime {}
find  -type f -print0 | xargs -0 -r -I {}  xattr -d user.rmlint.blake2b.cksum {}" 

can clean out xattr post manually in filer hierachy, but very slow...

After above 'cleaning' is possible to force rehashing files again (with -C, -U, --xattr*)  but after going to finish (or forced to stop with ctrl-C), is not possible to re-run again without 'cleaning' again and back to square one.

Is also risk for pollution in xattr ie. if try different hash-algoritm to force to recalculate hashes (yes is running in this case - one time...),  but not delete old (wrong) hashes and timestamp if using different hash algorithm,  and many file system have very limit space for storing xattr (~ 4 Kbyte)

**Comment by sahib** [comment]
> But unfortunately, I found more strange thing, is not tracked down yet and not always happen - but for large files, say > 30 GByte each seems checksum written to xattribute for files in directory long before is possible can be calculated (based of time needs for hashing for each file ), and also different value for exact same copy of files on different places - and worse, this cannot fix it with re-run again with -U, -C and -xattr* argument depend rmlint find files differ from reading of xattr in early stage (and seems have highest priority) and judge them as different and put out them from hashing later.
>
> [...]

Sorry, I had a hard time understand the issue... maybe I re-phrase what I think I understood:

- There were files that had the wrong checksum in xattrs, but I'm unsure how you checked this.
- You want to clear existing xattrs, but I don't know exactly what was wrong with them.
- Clearing them using ``--xattr=clear`` is indeed a bit cumbersome to use. (I agree)

A few questions on top of that:

- What was the ``rmlint`` ommand you used? Maybe explain the issue it along those commands.
- Do you have an example of a file that has the "wrong" checksum (only metadata)?

Also, possibly related: I pushed a small fix here 9e16c5ddd6ccf8c3b8bafb35344bc49304cd3a01 - before that ``rmlint`` would not take xattrs that had a comma in them - even though older versions of it created them like that. ``g_strtod`` can handle both cases.

**Comment by xxargs** [comment]
sorry for late response - it was a heavy work week and no energy to study the problem more thoroughly

I hope down is understandable after google translate...

example used on two snap-shot identical borg-backup archive with subdirectory '0'  
with 10000 files around 50 Mbyte each 

"rmlint  -v -C -U -g  -T df  -r  -b -e -m -k  /media/disk/old_snapshot/backup/data/0/ // media/disk/masterbackup/backup/data/0/"

Just at the beginning of the program run of rlint all files (2 * 10000 pcs) get xattr value with blake hash value and this value could possibly not have come from any calculation since it starts to be calculated only after all files have got their xattr value already.

when rmlint has done the hashing on all its files and start writing to the json file and rmlint.sh then the hash sum on xattr of all files is updated again to its correct value.


example down: short after start rmlint for file '1' give different blake hash for identical files in masterbackup and old_snapshot and seems randomized value  and impossible yet be calculated already based of time consumed for hashing:

user@/media/disk/masterbackup/backup/data/0# xattr -l 1 
btrfs.compression: zlib
user.rmlint.blake2b.mtime: 1476057826,000000000
user.rmlint.blake2b.cksum:
0000   32 35 61 66 66 66 30 35 30 39 32 30 30 31 34 62    25afff050920014b
0010   63 34 32 66 66 32 34 33 62 30 33 38 62 66 30 35    c42ff243b038bf05
0020   39 35 62 36 39 37 39 34 62 32 38 36 36 34 36 32    95b69794b2866462
0030   39 65 37 33 64 33 31 34 34 38 64 64 64 36 34 64    9e73d31448ddd64d
0040   36 66 38 31 33 38 64 32 66 36 33 38 61 36 63 66    6f8138d2f638a6cf
0050   34 39 36 36 32 32 65 37 66 64 66 38 39 39 32 36    496622e7fdf89926
0060   38 65 34 31 36 66 66 62 64 63 33 32 30 38 64 33    8e416ffbdc3208d3
0070   38 39 64 64 66 34 31 38 64 37 63 37 33 39 64 39    89ddf418d7c739d9
0080   00                                                 .

user@/media/disk/old_snapshot/backup/data/0# xattr -l 1 
btrfs.compression: zlib
user.rmlint.blake2b.mtime: 1476057826,000000000
user.rmlint.blake2b.cksum:
0000   39 66 63 31 39 32 30 61 39 34 63 39 32 32 61 65    9fc1920a94c922ae
0010   32 63 35 62 63 35 31 32 63 66 35 36 35 39 32 61    2c5bc512cf56592a
0020   37 39 38 63 66 62 34 37 31 37 39 32 33 38 38 37    798cfb4717923887
0030   31 34 63 62 38 65 35 31 38 38 34 64 64 36 33 36    14cb8e51884dd636
0040   36 36 66 65 62 64 65 62 62 66 35 32 61 62 33 32    66febdebbf52ab32
0050   35 37 31 32 61 65 64 30 34 34 62 30 30 36 66 64    5712aed044b006fd
0060   62 33 64 64 37 63 30 32 32 66 38 30 62 31 35 61    b3dd7c022f80b15a
0070   30 36 33 35 37 39 31 34 30 33 66 30 39 65 66 36    0635791403f09ef6
0080   00



After while and rmlint done hash calculation and writing out to json-file and rmlint.sh seems hash-value updated to correct value for both identical files and show: 

user@/media/disk/masterbackup/backup/data/0# xattr -l 1  
btrfs.compression: zlib
user.rmlint.blake2b.mtime: 1476057826,000000000
user.rmlint.blake2b.cksum:
0000   33 39 63 66 33 31 35 33 35 62 30 35 39 32 66 63    39cf31535b0592fc
0010   35 62 66 65 66 34 64 31 65 31 31 63 30 37 34 61    5bfef4d1e11c074a
0020   64 30 36 66 61 39 64 32 61 31 33 62 37 38 35 63    d06fa9d2a13b785c
0030   66 37 66 33 33 38 62 34 66 66 37 30 65 63 31 30    f7f338b4ff70ec10
0040   36 34 35 62 34 36 66 61 33 30 31 65 30 33 65 37    645b46fa301e03e7
0050   31 38 62 31 31 32 66 34 33 63 33 65 36 30 65 35    18b112f43c3e60e5
0060   64 64 34 65 64 63 62 32 30 66 64 30 62 36 66 30    dd4edcb20fd0b6f0
0070   64 63 66 64 31 31 32 65 35 65 39 63 66 66 64 61    dcfd112e5e9cffda
0080   00                                                 .


user@/media/disk/old_snapshot/backup/data/0# xattr -l 1
btrfs.compression: zlib
user.rmlint.blake2b.mtime: 1476057826,000000000
user.rmlint.blake2b.cksum:
0000   33 39 63 66 33 31 35 33 35 62 30 35 39 32 66 63    39cf31535b0592fc
0010   35 62 66 65 66 34 64 31 65 31 31 63 30 37 34 61    5bfef4d1e11c074a
0020   64 30 36 66 61 39 64 32 61 31 33 62 37 38 35 63    d06fa9d2a13b785c
0030   66 37 66 33 33 38 62 34 66 66 37 30 65 63 31 30    f7f338b4ff70ec10
0040   36 34 35 62 34 36 66 61 33 30 31 65 30 33 65 37    645b46fa301e03e7
0050   31 38 62 31 31 32 66 34 33 63 33 65 36 30 65 35    18b112f43c3e60e5
0060   64 64 34 65 64 63 62 32 30 66 64 30 62 36 66 30    dd4edcb20fd0b6f0
0070   64 63 66 64 31 31 32 65 35 65 39 63 66 66 64 61    dcfd112e5e9cffda
0080   00                                                 .

both have same blake-hash now.

the problem arises - which I did not think of before, is if you break the program execution of rmlint ctrl-C before it is complete - maybe to adjust the parameter or divide it into smaller items, because it became too much in a sweep

Then it has already put random blake hash in xattr on all file names already (which happens within the first minutes of the run) and at ctrl-C, maybe some of these files will be updated with the correct hash value while the rest will retain their old random value with no relation to the file.

when you do a new run with rmlint to take advantage of the stored hash values, rmlint will reading xattr and find the hash-value and this way  find many files without duplicate (because value was initiated with random value in the beginning and then not corrected with correct value late ) which means that many files that have duplicates, because the random hash sum is sorted out as unique.

this situation is also with xattr=clear that the files you want to clear / reset because they have received incorrect hash sum are already sorted in a previous step as a unique file in rmlint and you have a situation that you cannot delete the file names xattr-value that received incorrect hash sum.

I currently have about 2 million files with this incorrect hash sum just because I pressed ctrl-C at some point when rmlint  wasn't finish  and if should use 

"find -type f -print0 | xargs -0 -r -I {} xattr -d user.rmlint.blake2b.mtime {}
find -type f -print0 | xargs -0 -r -I {} xattr -d user.rmlint.blake2b.cksum {}"

to make it undone,  it will take weeks, yes maybe months when this instruction goes with about 1 file per second...

In that regard, it is important to get a function that unconditionally clears xattr on all rmlint data on all files it traverses through the file tree, whether there are duplicates or not, now it became a bit precarious mode that you can not easily fix xattr-value when it did not go as planned or because it happened to interrupt a running rmlint prematurely.

Also question why setting xattr with blake hash value already from beginning with random value - before valid blake-hash sum for the file exist... perhaps write ordering bug?

**Comment by xxargs** [comment]
after more running...

I becoming crazy...

same setup as latest post with around 20000 files in two directory for compare

If you look at the current file names in mastebackup directory with -xattr -l filname directly after starting rmlint (max 10 seconds later) you can see that not all 10000 files have their xattr added with random numbers yet, but 1 minute later have values ​​and when the directory is complete, it continues on the next old-snapshot directory it compares with.

It seems to be a process / thread completely runs wild and disconnected from the actual hash calculation and runs through the file tree as soon as it can and put xattr with fake hash without synchronizing with the time-consuming hash calculation which, when done again, adds the correct correct hash value later for the file name...

**Comment by sahib** [comment]
That's because the xattrs is [currently written on every hash iteration](https://github.com/sahib/rmlint/blob/master/lib/shredder.c#L690). In other words: there will be many intermediate (and thus "incorrect") checksums. I actually don't know why it's there and I suspect it's an oversight from the previous caching implementation. You have a point regarding writing xattrs once when the file finished hashing.

> this situation is also with xattr=clear that the files you want to clear / reset because they have received incorrect hash sum are already sorted in a previous step as a unique file in rmlint and you have a situation that you cannot delete the file names xattr-value that received incorrect hash sum.

That's actually not what the implementation does. The clearing happens early during [traversing files](https://github.com/sahib/rmlint/blob/master/lib/traverse.c#L215) - that still affects all files that are looked at, not only unique ones. That's only the theory, I didn't test it for some time and I can't remember if we have a unit test for that so I will test and see if it can be made more robust.

**Comment by sahib** [comment]
I implemented a fix as of e42de9c3eb53ad5c66a1cac5d951ae3cbd593148 - now every file should be written only once per ``rmlint`` run. Please try that out. Also finally got around to write a test for this in b52f87570d9ccb3ee67a6e1c373637cab853e8ac.
